File: Episode10-Powerups/Love2D/powerup.lua
local Class = require "lib.Class"
local Powerup = Class:derive("Powerup")
local Rectangle = require "lib.Rectangle"
local Circle = require "lib.Circle"
local Shared = require "shared"

function Powerup:new(spriteList, centre)
	--[[ class constructor, takes table of images
		 centre of the mob destroyed {x,y} 
		spriteList = {shield = <image>, gun = <image>}
	]]
	self.centre = centre
	self.spriteList = spriteList 	-- table of images
	-- choose random image
	local keys = {}
	for k,v in pairs(spriteList) do
		table.insert(keys, k)
	end
	-- keys = {"shield", "gun"}
	self.type = keys[math.random(1, #keys)] -- math.random returns 1 or 2, keys[1] returns 'gun'
	self.image = spriteList[self.type] 
	self.speedY = 400
	if Shared.debug then
		self.speedY = 200
	end
	self.active = true
	self.rect = Rectangle(centre[1] - (self.image:getWidth() / 2),
						  centre[2] - (self.image:getHeight() / 2),
						  self.image:getWidth(),
						  self.image:getHeight())
	self.circle = Circle(self.rect.centre.x, self.rect.centre.y, self.image:getWidth() / 2)
end

function Powerup:getCircle()
	return self.circle
end

function Powerup:getRect()
	return self.rect
end

function Powerup:update(dt)
	self.rect:updatePosition(0, self.speedY * dt)
	self.circle:update(self.rect.centre.x, self.rect.centre.y)
	-- check if Powerup has gone off bottom of screen
	if self.rect.top >Shared.HEIGHT then
		self.active = false
	end
	return self.active -- when false, powerup can be set to nil
end

function Powerup:draw()
	-- love.graphics.draw( drawable, x, y, r, sx, sy, ox, oy, kx, ky )
	if self.active then
		love.graphics.draw(self.image, self.rect.x, self.rect.y)
		if Shared.debug then
			love.graphics.setColor(Shared.RED)
			love.graphics.rectangle("line", self.rect.x, self.rect.y, self.rect.w, self.rect.h)
			love.graphics.setColor(Shared.WHITE)
		end
	end
end

return Powerup

File: Episode10-Powerups/Pygame/powerup.py
import pygame, random
import shared, circle

class Pow():
	def __init__(self, images, center):
		self.type = random.choice(['shield', 'gun'])
		self.images = images
		self.image = self.images[self.type]
		self.image.set_colorkey(shared.BLACK)	
		self.rect = self.image.get_rect()
		self.rect.center = center
		self.speedy = 500
		self.active = True
		self.circle = circle.Circle(self.rect.centerx, self.rect.centery, self.rect.width / 2)
		self.colour = shared.RED
	
	def get_rect(self):
		return self.rect

	def get_circle(self):
		self.circle.x = self.rect.centerx
		self.circle.y = self.rect.y	+ self.circle.radius	
		return self.circle		

	def update(self, dt):
		self.rect.y += self.speedy * dt
		# delete if moves off bottom of screen
		if self.rect.top > shared.HEIGHT:
			self.active = False
			
		return self.active
			
	def draw(self) -> None:
		shared.screen.blit(self.image, self.rect)
		if shared.debug:
			pygame.draw.circle(shared.screen, self.colour, self.circle.center, self.circle.radius, 1)	